File: Arboles/bfs/bfs-matriz-ady.cpp
#include <vector>
#include <map>
#include <string>
#include <cstdio>
#include <iostream>
#include <queue>

using namespace std;

int n;
int adj[1000][1000];
bool visitados[1000];

void bfsAux(int v){
  int u, i;
  queue<int> cola;
  visitados[v] = true;
  cola.push(v);

  while(!cola.empty()){
    u = cola.front();
    printf("%d\n", u);
    cola.pop();

    for(i = 0; i < n; ++i){
      if(adj[u][i] && !visitados[i]){
	visitados[i] = true;
	cola.push(i);
      }
    }
  }
}

void bfs(){
  int i;

  //fill(visitados, visitados + n, false);
  for(i = 0; i < n; ++i)
    visitados[i] = false;

  for(i = 0; i < n; ++i){
    if(!visitados[i])
      bfsAux(i);
  }
}


int main(){
  int m, u, v, i, j;
  cin >> n >> m;

  for(i = 0; i < n; ++i){
    cin >> u;
    for(j = 0; j < u; ++j){
      cin >> v;
      adj[i][v] = 1;
    }
  }

  cout << "Grafo" << endl;

  for(i = 0; i < n; ++i){
    for(j = 0; j < n; ++j)
      cout << adj[i][j] << " ";
    cout << endl;
  }

  return 0;
}

File: Arboles/bfs/bfs-matriz-ady.py
from sys import stdin
from collections import deque

MAX, n = 1000, int()
adj = [[0 for _ in range(MAX)] for _ in range(MAX)]
visitados = [False for _ in range(MAX)]
#p = [-1 for _ in range(MAX)]

def bfsAux(v):
    cola = deque()
    cola.append(v)
    visitados[v] = True

    while len(cola) != 0:
        u = cola.popleft()
        print(u)

        for i in range(n):
            if adj[u][i] == 1 and not visitados[i]:
                cola.append(i)
                visitados[i] = True
                #p[i] = v

def bfs():
    for i in range(n):
        visitados[i] = False
        #p[i] = -1
    for i in range(n):
        if not visitados[i]:
            bfsAux(i)

def main():
    global n, visitados, adj
    n, m = list(map(int, stdin.readline().split()))

    for i in range(n):
        vals = list(map(int, stdin.readline().split()))
        u = vals[0]
        for j in range(u):
            v = vals[j + 1]
            adj[i][v] = 1

    print("Grxafo")
    for i in range(n):
        for j in range(n):
            print("%d" % adj[i][j], end = ' ')
        print("")

    print("BFS")
    bfs()
    for i in range(n):
        print(p[i], end = ' ')
    print("")

main()
